src/Microsoft.OData.Core/Evaluation/ODataResourceMetadataContext.cs

```diff
@@ -288,13 +288,19 @@ private static void ValidateEntityTypeHasKeyProperties(KeyValuePair<string, obje
             Debug.Assert(resource != null, "resource != null");
             Debug.Assert(propertyKind == ODataPropertyKind.Key || propertyKind == ODataPropertyKind.ETag, "propertyKind == ODataPropertyKind.Key || propertyKind == ODataPropertyKind.ETag");
 
-            KeyValuePair<string, object>[] properties = EmptyProperties;
+            List<KeyValuePair<string, object>> properties = new List<KeyValuePair<string, object>>(); ;
             if (resource.NonComputedProperties != null)
             {
-                properties = resource.NonComputedProperties.Where(p => p.SerializationInfo != null && p.SerializationInfo.PropertyKind == propertyKind).Select(p => new KeyValuePair<string, object>(p.Name, GetPrimitiveOrEnumPropertyValue(actualEntityTypeName, p, propertyKind == ODataPropertyKind.Key))).ToArray();
+                foreach(ODataProperty property in resource.NonComputedProperties)
+                {
+                    if(property.SerializationInfo != null && property.SerializationInfo.PropertyKind == propertyKind)
+                    {
+                        properties.Add(new KeyValuePair<string, object>(property.Name, GetPrimitiveOrEnumPropertyValue(actualEntityTypeName, property, propertyKind == ODataPropertyKind.Key)));
+                    }
+                }                
             }
 
-            return properties;
+            return properties.ToArray();

```

Comment:
>.ToArray() [](start = 29, length = 10)

How is this used? Is it more efficient to return it as a list, rather than copy to an array?

Reply:
In other methods the returned key property is accessed, and passed down to other methods as array (in most cases) . Thats the reason we returned it as array

---
In reply to: [600680945](https://github.com/OData/odata.net/pull/2024#discussion_r600680945) [](ancestors = 600680945)